core/trino-spi/src/main/java/io/trino/spi/statistics/StatsUtil.java

```diff
@@ -85,4 +85,44 @@ public static OptionalDouble toStatsRepresentation(Type type, Object value)
 
         return OptionalDouble.empty();
     }
+
+    public static boolean areStatsRepresentationsCompatible(Type type, Type otherType)
+    {
+        requireNonNull(type, "type is null");
+        requireNonNull(otherType, "otherType is null");
+
+        if (type.equals(otherType)) {
+            return true;
+        }
+
+        if (type == BOOLEAN || type == REAL || type == DOUBLE || type == DATE) {
+            return false;
+        }
+        if (type == TINYINT || type == SMALLINT || type == INTEGER || type == BIGINT) {
+            return (otherType == TINYINT || otherType == SMALLINT || otherType == INTEGER || otherType == BIGINT);
+        }
+        if (type instanceof DecimalType) {
+            if (!(otherType instanceof DecimalType)) {
+                return false;
+            }
+            DecimalType decimalType = (DecimalType) type;
+            DecimalType otherDecimalType = (DecimalType) otherType;
+            return (decimalType.isShort() == otherDecimalType.isShort())
+                    && (decimalType.getScale() == otherDecimalType.getScale());
```

Comment:
For some definitions of "compatible", this `decimal(p, s)` and `decimal(p+1, s+1)` are "compatible"

Reply:
I've assumed that stats compatibility between two types means that `toStatsRepresentation` returns the same result for a value with either Type. Please let me know if that definition is inaccurate.
A given input `value` can either be cast to a `long` for the "short" types or to a `Int128/LongTimestamp/LongTimestampWithTimeZone` for the "long" types in `toStatsRepresentation`. That's why I kept the conditions for both short or both long type and same scale.

Reply:
> I've assumed that stats compatibility

Well, this shouldn't be "assumed". It should be "defined", as it's being added in this PR :) 

Anyway, whatever the definition is currently, it's not quite coherent yet. 
For example, you say that `integer` and `bigint` are compatible. `bigint` can represent all values of `integer` type and some more -- it's strictly wider and for the overlapping values, their stats representation is the same.

Following this example, `decimal(p, s)` and `decimal(p+1, s)` should be compatible -- the second one is strictly wider and for the overlapping values, their stats representation is the same.

TL;DR whatever the definition is going to be, coherence demands that there is no short/short or long/long condition here.